File: packages/checkin-core/src/token.ts
import { createHmac, timingSafeEqual } from "node:crypto";

/**
 * Pass token.
 *
 * Format:  <pass>.<event>.<v>.<sig>
 *
 * UUIDs are packed to 16 raw bytes and base64url'd to 22 characters rather
 * than carried as 36-character hex. The whole token lands around 62 chars,
 * which keeps the QR at a low version — fewer, larger modules, and a much
 * better scan in a dark hall on a cracked screen.
 *
 * The signature is HMAC-SHA256 truncated to 12 bytes. Forging one is 2^96
 * work, and the server re-derives every outcome on sync anyway, so the
 * token only has to be good enough to trust for the few seconds between
 * the scan and the guest walking through.
 *
 * No personal data is in here. A leaked token reveals two opaque ids.
 */

export type EventKey = {
  eventId: string;
  eventName: string;
  tokenVersion: number;
  key: Buffer;
};

export type TokenPayload = {
  passId: string;
  eventId: string;
  tokenVersion: number;
};

const UUID_RE =
  /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

function packUuid(uuid: string): string {
  if (!UUID_RE.test(uuid)) throw new Error(`not a uuid: ${uuid}`);
  return Buffer.from(uuid.replace(/-/g, ""), "hex").toString("base64url");
}

function unpackUuid(packed: string): string {
  const b = Buffer.from(packed, "base64url");
  if (b.length !== 16) throw new Error("bad uuid packing");
  const h = b.toString("hex");
  return `${h.slice(0, 8)}-${h.slice(8, 12)}-${h.slice(12, 16)}-${h.slice(
    16,
    20,
  )}-${h.slice(20)}`;
}

function sign(body: string, key: Buffer): string {
  return createHmac("sha256", key).update(body).digest().subarray(0, 12)
    .toString("base64url");
}

export function issueToken(p: TokenPayload, key: Buffer): string {
  const body = `${packUuid(p.passId)}.${packUuid(p.eventId)}.${p.tokenVersion}`;
  return `${body}.${sign(body, key)}`;
}

export type VerifyOk = {
  ok: true;
  payload: TokenPayload;
  /** Which held key verified it — may not be the event being scanned. */
  matched: EventKey;
};
export type VerifyFail = { ok: false; reason: "malformed" | "no_key" | "stale_version" };

/**
 * Tries every key the device holds, not just the current event's.
 *
 * That is deliberate. With one key you cannot tell a forgery from last
 * week's wedding, and the usher gets "invalid" for a guest holding a
 * perfectly genuine pass to a different event. With all of them you can
 * say which event it belongs to, which is the difference between a useful
 * message and a shrug.
 */
export function verifyToken(raw: string, keys: EventKey[]): VerifyOk | VerifyFail {
  const parts = raw.trim().split(".");
  if (parts.length !== 4) return { ok: false, reason: "malformed" };

  // Length checked above; assertion is type-level only.
  const [pass, event, verStr, sig] = parts as [string, string, string, string];
  const body = `${pass}.${event}.${verStr}`;

  let payload: TokenPayload;
  try {
    payload = {
      passId: unpackUuid(pass),
      eventId: unpackUuid(event),
      tokenVersion: Number(verStr),
    };
  } catch {
    return { ok: false, reason: "malformed" };
  }
  if (!Number.isInteger(payload.tokenVersion)) {
    return { ok: false, reason: "malformed" };
  }

  const given = Buffer.from(sig, "base64url");
  if (given.length !== 12) return { ok: false, reason: "malformed" };

  for (const k of keys) {
    const expected = Buffer.from(sign(body, k.key), "base64url");
    if (expected.length === given.length && timingSafeEqual(expected, given)) {
      // A pass reissue bumps token_version. Old codes stop working without
      // needing every one of them on a revocation list.
      if (payload.tokenVersion !== k.tokenVersion) {
        return { ok: false, reason: "stale_version" };
      }
      return { ok: true, payload, matched: k };
    }
  }
  return { ok: false, reason: "no_key" };
}

File: apps/scanner/lib/core/token.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:crypto/crypto.dart';

/// Pass token — Dart port of packages/checkin-core/src/token.ts.
///
/// This file and checkin.dart are the ONLY logic duplicated across
/// languages (the TS original's own rule). Port behaviour, not style;
/// the test cases ported with it are the specification.
///
/// Format:  `<pass>.<event>.<v>.<sig>`
/// UUIDs packed to 16 raw bytes, base64url (no padding), ~62 chars total.
/// Signature is HMAC-SHA256 truncated to 12 bytes.

class EventKey {
  final String eventId;
  final String eventName;
  final int tokenVersion;
  final List<int> key;
  const EventKey({
    required this.eventId,
    required this.eventName,
    required this.tokenVersion,
    required this.key,
  });
}

class TokenPayload {
  final String passId;
  final String eventId;
  final int tokenVersion;
  const TokenPayload({
    required this.passId,
    required this.eventId,
    required this.tokenVersion,
  });
}

sealed class VerifyResult {
  const VerifyResult();
  bool get ok => this is VerifyOk;
}

class VerifyOk extends VerifyResult {
  final TokenPayload payload;

  /// Which held key verified it — may not be the event being scanned.
  final EventKey matched;
  const VerifyOk({required this.payload, required this.matched});
}

/// reason: 'malformed' | 'no_key' | 'stale_version'
class VerifyFail extends VerifyResult {
  final String reason;
  const VerifyFail(this.reason);
}

final _uuidRe = RegExp(
  r'^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$',
  caseSensitive: false,
);

String _b64url(List<int> bytes) => base64Url.encode(bytes).replaceAll('=', '');

Uint8List _unb64url(String s) => base64Url.decode(base64Url.normalize(s));

String packUuid(String uuid) {
  if (!_uuidRe.hasMatch(uuid)) {
    throw ArgumentError('not a uuid: $uuid');
  }
  final hex = uuid.replaceAll('-', '').toLowerCase();
  final bytes = Uint8List(16);
  for (var i = 0; i < 16; i++) {
    bytes[i] = int.parse(hex.substring(i * 2, i * 2 + 2), radix: 16);
  }
  return _b64url(bytes);
}

String unpackUuid(String packed) {
  final b = _unb64url(packed);
  if (b.length != 16) throw ArgumentError('bad uuid packing');
  final h = b.map((x) => x.toRadixString(16).padLeft(2, '0')).join();
  return '${h.substring(0, 8)}-${h.substring(8, 12)}-${h.substring(12, 16)}-'
      '${h.substring(16, 20)}-${h.substring(20)}';
}

String _sign(String body, List<int> key) {
  final mac = Hmac(sha256, key).convert(utf8.encode(body));
  return _b64url(mac.bytes.sublist(0, 12));
}

String issueToken(TokenPayload p, List<int> key) {
  final body = '${packUuid(p.passId)}.${packUuid(p.eventId)}.${p.tokenVersion}';
  return '$body.${_sign(body, key)}';
}

bool _constantTimeEquals(List<int> a, List<int> b) {
  if (a.length != b.length) return false;
  var diff = 0;
  for (var i = 0; i < a.length; i++) {
    diff |= a[i] ^ b[i];
  }
  return diff == 0;
}

/// Tries every key the device holds, not just the current event's —
/// that is what turns "invalid" into "this pass is for Yusuf & Maryam".
VerifyResult verifyToken(String raw, List<EventKey> keys) {
  final parts = raw.trim().split('.');
  if (parts.length != 4) return const VerifyFail('malformed');

  final pass = parts[0], event = parts[1], verStr = parts[2], sig = parts[3];
  final body = '$pass.$event.$verStr';

  final TokenPayload payload;
  try {
    final version = int.parse(verStr);
    payload = TokenPayload(
      passId: unpackUuid(pass),
      eventId: unpackUuid(event),
      tokenVersion: version,
    );
  } catch (_) {
    return const VerifyFail('malformed');
  }

  final Uint8List given;
  try {
    given = _unb64url(sig);
  } catch (_) {
    return const VerifyFail('malformed');
  }
  if (given.length != 12) return const VerifyFail('malformed');

  for (final k in keys) {
    final expected = _unb64url(_sign(body, k.key));
    if (_constantTimeEquals(expected, given)) {
      // A pass reissue bumps token_version. Old codes stop working without
      // needing every one of them on a revocation list.
      if (payload.tokenVersion != k.tokenVersion) {
        return const VerifyFail('stale_version');
      }
      return VerifyOk(payload: payload, matched: k);
    }
  }
  return const VerifyFail('no_key');
}
